### PR TITLE
Revisit versioner middlewares

### PR DESCRIPTION
**Pull request by ericproulx**
This PR is a refactor of all Grape::Middleware::Versioner

First, they all use the `default_options` like the other middlewares. Since default `version_options` is including another Hash, I had to replace `merge` by `deep_merge` in Grape::Middleware::Base.

Second, the `using` strategy is now [dynamically](https://github.com/ruby-grape/grape/blob/c2f3578b52871eb78d87eaeb9255a8b782d90a03/lib/grape/middleware/versioner.rb#L19) solved like validators. It's all based on the name of the class.

Third, I've added a `versioner_helpers` that all Grape::Middleware::Versioner include. At least, all the `options[...]` are exposed through an helper function which reduces the `[]` calls.

### CODE REVIEW

File: lib/grape/middleware/versioner.rb [hunk 1]
```diff
@@ -13,21 +13,12 @@ module Middleware
...
-          Header
-        when :param
-          Param
-        when :accept_version_header
-          AcceptVersionHeader
-        else
-          raise Grape::Exceptions::InvalidVersionerOption.new(strategy)
-        end
+        Grape::Middleware::Versioner.const_get(:"#{strategy.to_s.classify}")
+      rescue NameError
```
Comment (dblock) [inline lib/grape/middleware/versioner.rb:20]: I don't like that we rely on an exception here. It also allows you to register a strategy without anyone noticing, which ends up being used as an API.

File: lib/grape/middleware/versioner/accept_version_header.rb [hunk 2]
```diff
@@ -17,45 +17,22 @@ module Versioner
...
-          return if potential_version.empty?
+        def before
+          potential_version = env[Grape::Http::Headers::HTTP_ACCEPT_VERSION]&.strip
+          throw_error_not_acceptable('Accept-Version header must be set.') if strict? && potential_version.blank?
 
-          # If the requested version is not supported:
-          throw :error, status: 406, headers: error_headers, message: 'The requested version is not supported.' unless versions.any? { |v| v.to_s == potential_version }
+          return if potential_version.blank?
 
+          throw_error_not_acceptable('The requested version is not supported.') unless potential_version_match?(potential_version)
```
Comment (dblock) [inline lib/grape/middleware/versioner/accept_version_header.rb:28]: Probably `not_acceptable!` makes more sense (needs a bang and is similar to other throws).

File: lib/grape/middleware/versioner/param.rb [hunk 3]
```diff
@@ -19,31 +19,15 @@ module Versioner
...
+        include VersionerHelpers
 
         def before
-          potential_version = Rack::Utils.parse_nested_query(env[Rack::QUERY_STRING])[paramkey]
-          return if potential_version.nil?
+          potential_version = Rack::Utils.parse_nested_query(env[Rack::QUERY_STRING])[parameter_key]
+          return if potential_version.blank?
 
-          throw :error, status: 404, message: '404 API Version Not Found', headers: { Grape::Http::Headers::X_CASCADE => 'pass' } if options[:versions] && !options[:versions].find { |v| v.to_s == potential_version }
+          throw_api_version_not_found unless potential_version_match?(potential_version)
```
Comment (dblock) [inline lib/grape/middleware/versioner/param.rb:28]: version_not_found!

**Review by dblock — COMMENTED** [review]
Comments below. 

Beef up the description of the PR a bit to explain why this is better. Why is it? :)

File: lib/grape/middleware/versioner.rb (hunk 1 above)
Comment (ericproulx) [reply lib/grape/middleware/versioner.rb:20]: I do understand your point about registering without anyone noticing but on the other hand we're already embracing the pattern through [coercing](https://github.com/ruby-grape/grape/blob/2b8567a1ca49b3d0236fa82deb2edef34c8839cf/lib/grape/validations/types/primitive_coercer.rb#L35) and [validators](https://github.com/ruby-grape/grape/blob/2b8567a1ca49b3d0236fa82deb2edef34c8839cf/lib/grape/validations.rb#L25). Now, everyone can build their own `versioner` and they just need to follow the naming convention like the validators.

Comment (ericproulx) [reply lib/grape/middleware/versioner.rb:20]: In the future we could imagine having a separate gem just for versioning. It's a bit like how [Faraday](https://github.com/lostisland/faraday/blob/main/UPGRADING.md#adapters-have-moved) evolved through time by moving out the adapters. I know versioning ain't that big compared to adapters but its the idea

Comment (dblock) [reply lib/grape/middleware/versioner.rb:20]: Ok, I buy the argument of we're already doing it.

**Review by dblock — COMMENTED** [review]
The entire spec/grape/util/accept_header_handler_spec.rb is gone. Where ele do we cover the accept version negotiation and x-cascade handling for that one in tests?

**Comment by ericproulx** [comment]
> The entire spec/grape/util/accept_header_handler_spec.rb is gone. Where ele do we cover the accept version negotiation and x-cascade handling for that one in tests?

There's already the accept_version_header_spec that covers all cases. Coverage is 100%

**Pull request merged**